File: Java/27_Trie/H1707. Maximum XOR With an Element From Array.java
// 1707. Maximum XOR With an Element From Array
// https://leetcode.com/problems/maximum-xor-with-an-element-from-array/

class Solution {
    // Bitwise trie node for integers
    private static class Node {
        Node[] child = new Node[2];
    }

    private void insert(Node root, int num) {
        Node cur = root;
        for (int bit = 31; bit >= 0; bit--) {
            int b = (num >> bit) & 1;
            if (cur.child[b] == null) cur.child[b] = new Node();
            cur = cur.child[b];
        }
    }

    private int query(Node root, int num) {
        Node cur = root;
        int ans = 0;
        for (int bit = 31; bit >= 0; bit--) {
            int b = (num >> bit) & 1;
            int want = 1 - b;
            if (cur.child[want] != null) {
                ans |= (1 << bit);
                cur = cur.child[want];
            } else if (cur.child[b] != null) {
                cur = cur.child[b];
            } else {
                break;
            }
        }
        return ans;
    }

    public int[] maximizeXor(int[] nums, int[][] queries) {
        java.util.Arrays.sort(nums);

        int qn = queries.length;
        int[][] qs = new int[qn][3]; // xi, mi, originalIndex
        for (int i = 0; i < qn; i++) {
            qs[i][0] = queries[i][0];
            qs[i][1] = queries[i][1];
            qs[i][2] = i;
        }

        java.util.Arrays.sort(qs, new java.util.Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                return Integer.compare(a[1], b[1]); // sort by mi
            }
        });

        int[] ans = new int[qn];
        java.util.Arrays.fill(ans, -1);

        Node root = new Node();
        int idx = 0, n = nums.length;

        for (int[] q : qs) {
            int x = q[0], m = q[1], qi = q[2];

            // Insert all nums <= m into trie
            while (idx < n && nums[idx] <= m) {
                insert(root, nums[idx]);
                idx++;
            }

            if (idx == 0) {
                ans[qi] = -1; // no valid element
            } else {
                ans[qi] = query(root, x);
            }
        }

        return ans;
    }

    public static void main(String[] args){
        
        Solution sol = new Solution();

        int[] nums = {0,1,2,3,4};
        int[][] queries = {{3,1},{1,3},{5,6}};

        int[] ans = sol.maximizeXor(nums, queries);

        for (int i = 0; i < ans.length; i++) {
            System.out.println(ans[i]);
        }
    }
}

File: JavaScript/27_Trie/H1707. Maximum XOR With an Element From Array.js
/**
 * 1707. Maximum XOR With an Element From Array
 * https://leetcode.com/problems/maximum-xor-with-an-element-from-array/
 *
 * For each query [xi, mi], find max XOR of xi with element <= mi from nums.
 *
 * Type             | Details
 * -----------------+----------------------------------------------------
 * Time Complexity  | O((N + Q) * 32) – Process each number in trie
 * Space Complexity | O(N * 32) – Trie nodes
 */

function createTrieNode() {
    const children = [null, null]; // 0 and 1

    return {
        getChild(bit) { return children[bit]; },
        setChild(bit, node) { children[bit] = node; },
        hasChild(bit) { return children[bit] !== null; }
    };
}

function createTrie() {
    const root = createTrieNode();

    function insert(num) {
        let node = root;
        for (let i = 31; i >= 0; i--) {
            const bit = (num >> i) & 1;
            if (!node.hasChild(bit)) {
                node.setChild(bit, createTrieNode());
            }
            node = node.getChild(bit);
        }
    }

    function getMaxXOR(num) {
        if (!root.hasChild(0) && !root.hasChild(1)) {
            return -1;
        }

        let node = root;
        let maxXOR = 0;

        for (let i = 31; i >= 0; i--) {
            const bit = (num >> i) & 1;
            const oppositeBit = 1 - bit;

            if (node.hasChild(oppositeBit)) {
                maxXOR |= (1 << i);
                node = node.getChild(oppositeBit);
            } else if (node.hasChild(bit)) {
                node = node.getChild(bit);
            } else {
                return -1;
            }
        }

        return maxXOR;
    }

    return { insert, getMaxXOR };
}

/**
 * Maximum XOR With an Element From Array
 * @param {number[]} nums
 * @param {number[][]} queries
 * @returns {number[]}
 */
function maximizeXor(nums, queries) {
    // Sort nums
    nums.sort((a, b) => a - b);

    // Add index to queries and sort by mi
    const indexedQueries = queries.map((q, i) => [...q, i]);
    indexedQueries.sort((a, b) => a[1] - b[1]);

    const result = new Array(queries.length);
    const trie = createTrie();
    let numIndex = 0;

    for (const [xi, mi, queryIndex] of indexedQueries) {
        // Add all nums <= mi to trie
        while (numIndex < nums.length && nums[numIndex] <= mi) {
            trie.insert(nums[numIndex]);
            numIndex++;
        }

        result[queryIndex] = trie.getMaxXOR(xi);
    }

    return result;
}

// Test cases
const nums1 = [0, 1, 2, 3, 4];
const queries1 = [[3, 1], [1, 3], [5, 6]];
console.log("maximizeXor([0,1,2,3,4], [[3,1],[1,3],[5,6]]):", 
    maximizeXor(nums1, queries1)); // [3, 3, 7]

const nums2 = [5, 2, 4, 6, 6, 3];
const queries2 = [[12, 4], [8, 1], [6, 3]];
console.log("maximizeXor([5,2,4,6,6,3], [[12,4],[8,1],[6,3]]):", 
    maximizeXor(nums2, queries2)); // [15, -1, 5]
